numpyro/distributions/directional.py

```diff
@@ -0,0 +1,120 @@
+import jax.numpy as np
+from jax import lax, random
+
+from numpyro.distributions import constraints
+from numpyro.distributions.distribution import Distribution
+from numpyro.distributions.util import validate_sample
+from numpyro.util import copy_docs_from
+
+
+@copy_docs_from(Distribution)
+class VonMises(Distribution):
+    arg_constraints = {'location': constraints.real, 'concentration': constraints.positive}
+    support = constraints.real
+    s_cutoff_map = {np.dtype(np.float16): 1.8e-1,
+                    np.dtype(np.float32): 2e-2,
+                    np.dtype(np.float64): 1.2e-4}
+
+    def __init__(self, location, concentration, validate_args=None):
+        # von Mises doesn't work for concentration=0, so set to tiny value
+        try:
+            dtype = np.dtype(concentration.dtype)
+        except AttributeError:
+            dtype = np.dtype(type(concentration))
+        concentration = np.maximum(concentration, np.finfo(dtype).tiny)
+        self.dtype = dtype
+        self._concentration = concentration
+        self._loc = location
+
+        batch_shape = lax.broadcast_shapes(np.shape(concentration), np.shape(location))
+
+        super(VonMises, self).__init__(batch_shape=batch_shape,
+                                       validate_args=validate_args)
+
+    def _sample_centered(self, sample_shape, rng_key):
+        """ Compute centered von Mises samples using rejection sampling from [1] with wrapped Cauchy proposal.
+
+            *** References ***
+            [1] Luc Devroye "Non-Uniform Random Variate Generation", Springer-Verlag, 1986;
+                Chapter 9, p. 473-476. http://www.nrbook.com/devroye/Devroye_files/chapter_nine.pdf
+
+            :param sample_shape: shape of samples
+            :param rng_key: random number generator key
+            :return: centered samples from von Mises
+        """
+        conc = self._concentration
+
+        r = 1. + np.sqrt(1. + 4. * conc ** 2)
+        rho = (r - np.sqrt(2. * r)) / (2. * conc)
+        s_exact = (1. + rho ** 2) / (2. * rho)
+
+        s_approximate = 1. / conc
+
+        s_cutoff = VonMises.s_cutoff_map.get(self.dtype)
+
+        s = np.where(conc > s_cutoff, s_exact, s_approximate)
+
+        def body_function(i, *args):
+            done, _, w = args[0]
+            nonlocal rng_key
```

Comment:
Adding rng_key to args raise inconsistent dimensions exception in the loop.

Reply:
The error is a bit strange. How about using the following pattern
```python
def cond_fn(val):
    i, _, done, _, _ = val
    return (i < 100) & (~np.all(done))

def body_fn(val):
    i, rng_key, done, _, w = val
    ...
    return i + 1, rng_key, accept | done, u, w

_, _, _, u, w = lax.while_loop(cond_fn, body_fn, (np.array(0), rng_key, init_done, init_u, init_w))
```
(there might be some bugs but I think this logic should work - feel free to add a TODO: use while_loop instead of fori_loop if this still triggers errors for you)